### PR TITLE
Large Send Offload for UDP protocol.

### PR DESCRIPTION
**Pull request by yazimut**
Added support for LSO for UDP protocol.

* Issues fixed (if any):
   ** Alignment in some functions.

### CODE REVIEW

File: src/iperf_api.c [hunk 1]
```diff
@@ -104,6 +105,8 @@
     #define UDP_SEGMENT 103 /* Set GSO segmentation size */
 #endif /* UDP_SEGMENT */
 
+#define LSO_UDP_MIN_SEGSIZE 1024
```
Comment (cail) [inline src/iperf_api.c]: Why? thats not true! any size, including 1 byte works!

File: src/iperf_udp.c [hunk 2]
```diff
@@ -567,6 +643,26 @@ iperf_udp_connect(struct iperf_test *test)
 
     if (test->settings->varlen)
         srand(time(NULL));
+        
+    if (test->settings->lso_udp) {
+        int level, name, val;
+
+        if (test->settings->domain == PF_UNSPEC ||
+            test->settings->domain == PF_INET) {
+                level = SOL_IP;
```
Comment (cail) [inline src/iperf_udp.c:669]: why need this?

File: src/iperf_udp.c [hunk 3]
```diff
@@ -98,9 +106,11 @@ iperf_udp_recv(struct iperf_stream *sp)
 	    memcpy(&sec, sp->buffer, sizeof(sec));
 	    memcpy(&usec, sp->buffer+4, sizeof(usec));
 	    memcpy(&pc, sp->buffer+8, sizeof(pc));
+        memcpy(&lso_part, sp->buffer+12, sizeof(lso_part));
```
Comment (cail) [inline src/iperf_udp.c:109]: alignment. check other lines as well.

File: src/iperf_api.c [hunk 4]
```diff
@@ -1379,6 +1382,51 @@ iperf_parse_arguments(struct iperf_test *test, int argc, char **argv)
         warning("Debug output (-d) may interfere with JSON output (-J)");
     }
 
+    if (test->settings->lso_udp) {
```
Comment (cail) [inline src/iperf_api.c:1383]: not all linuxes support GSO_UDP socket option. We should somehow query that before use.

**Review by cail — CHANGES_REQUESTED** [review]

File: src/iperf_api.c (hunk 4 above)
Comment (yazimut) [reply src/iperf_api.c:1383]: I've made a check for supporting (iperf_api.c - line 98). But it works only at compile time. In runtime, using of UDP LSO will cause the error in "sendmsg" call - EINVAL.

File: src/iperf_udp.c (hunk 3 above)
Comment (yazimut) [reply src/iperf_udp.c:109]: Fixed

File: docs/invoking.rst [hunk 5]
```diff
@@ -399,10 +399,17 @@ the executable.
...
                  test. Tests are described in a JSON format.
+
           --varlen
                  enable variable random UDP packet size. Client simulates realis-
                  tic network load. Works only with -u key. It is recommended to 
                  disable speed limit with -b 0 key.
+
+          --lso-udp n
+                 enables Large Segmentation Offload for UDP protocol. "n" is a
+                 segment size, it cannot be less 1024 bytes or greater than MTU.
```
Comment (cail) [inline docs/invoking.rst:410]: remove notice on 1024b limit.

File: src/iperf_api.c [hunk 6]
```diff
@@ -90,6 +95,16 @@
 #include <pthread.h>
 #include <sys/sysinfo.h>
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 18, 0)
+    #error Your configuration isn't supporting LSO dor UDP! If you want to use it on\
```
Comment (cail) [inline src/iperf_api.c:99]: thats wrong. this way you disable iperf builds on all linuxes < 4.18.
thats wrong.

File: src/iperf_api.c [hunk 7]
```diff
@@ -804,25 +819,26 @@ iperf_parse_arguments(struct iperf_test *test, int argc, char **argv)
         {"nstreams", required_argument, NULL, OPT_NUMSTREAMS},
         {"xbind", required_argument, NULL, 'X'},
 #endif
-	{"pidfile", required_argument, NULL, 'I'},
```
Comment (cail) [inline src/iperf_api.c:807]: you should not touch other code formatting. please check your editor (probably autoformat was applied).
such changes are unrelated to our pull request.

File: src/iperf_api.c (hunk 4 above)
Comment (cail) [reply src/iperf_api.c:1383]: think that makes sense. EINVAL is enouth

File: src/iperf_udp.c (hunk 3 above)
Comment (cail) [reply src/iperf_udp.c:109]: but all other places??

File: src/iperf_udp.c [hunk 8]
```diff
@@ -87,20 +93,24 @@ iperf_udp_recv(struct iperf_stream *sp)
...
-	    memcpy(&usec, sp->buffer+4, sizeof(usec));
-	    memcpy(&pc, sp->buffer+8, sizeof(pc));
+        memcpy(&sec, sp->buffer, sizeof(sec));
+        memcpy(&usec, sp->buffer+4, sizeof(usec));
+        memcpy(&pc, sp->buffer+8, sizeof(pc));
+        memcpy(&lso_part, sp->buffer+12, sizeof(lso_part));
 	    sec = ntohl(sec);
 	    usec = ntohl(usec);
 	    pcount = ntohl(pc);
+        lso_part = ntohs(lso_part);
```
Comment (cail) [inline src/iperf_udp.c:113]: here

File: src/iperf_udp.c [hunk 9]
```diff
@@ -214,31 +228,102 @@ iperf_udp_send(struct iperf_stream *sp)
 
 	uint32_t  sec, usec;
 	uint64_t  pcount;
+    uint16_t  lso_part = 0;
```
Comment (cail) [inline src/iperf_udp.c:231]: here. always review your pull request by yourself first...

File: src/iperf_udp.c [hunk 10]
```diff
@@ -214,31 +228,102 @@ iperf_udp_send(struct iperf_stream *sp)
...
 
 	sec = htonl(before.secs);
 	usec = htonl(before.usecs);
 	pcount = htobe64(sp->packet_count);
-	
-	memcpy(sp->buffer, &sec, sizeof(sec));
-	memcpy(sp->buffer+4, &usec, sizeof(usec));
-	memcpy(sp->buffer+8, &pcount, sizeof(pcount));
+    lso_part = htons(lso_part);
```
Comment (cail) [inline src/iperf_udp.c:236]: here

File: src/iperf_udp.c [hunk 11]
```diff
@@ -567,6 +652,34 @@ iperf_udp_connect(struct iperf_test *test)
...
+                level = SOL_IP;
+                name  = IP_MTU_DISCOVER;
+                val   = IP_PMTUDISC_DO;
+        } else if (test->settings->domain == PF_INET6) {
+                level = SOL_IPV6;
+                name  = IPV6_MTU_DISCOVER;
+                val   = IPV6_PMTUDISC_DO;
+        }
+
+        if (setsockopt(s, level, name, &val, sizeof(val))) {
```
Comment (cail) [inline src/iperf_udp.c:678]: again, why you've added this? why you need MTU_DISCOVER for lso_udp specifically?

**Review by cail — CHANGES_REQUESTED** [review]
